drivers/gs.go

```diff
@@ -0,0 +1,162 @@
+package drivers
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/livepeer/go-livepeer/net"
+)
+
+type (
+	gsKeyJSON struct {
+		Type                    string `json:"type,omitempty"`
+		ProjectID               string `json:"project_id,omitempty"`
+		PrivateKeyID            string `json:"private_key_id,omitempty"`
+		PrivateKey              string `json:"private_key,omitempty"`
+		ClientEmail             string `json:"client_email,omitempty"`
+		ClientID                string `json:"client_id,omitempty"`
+		AuthURI                 string `json:"auth_uri,omitempty"`
+		TokenURI                string `json:"token_uri,omitempty"`
+		AuthProviderX509CertURL string `json:"auth_provider_x509_cert_url,omitempty"`
+		ClientX509CertURL       string `json:"client_x509_cert_url,omitempty"`
+	}
+
+	gsSigner struct {
+		jsKey     *gsKeyJSON
+		parsedKey *rsa.PrivateKey
+	}
+
+	gsOS struct {
+		s3OS
```

Comment:
This might feel nitpicky, but I think we can do better at naming here.  Should we call this `cloudOS` instead of `s3OS`, move it into `drivers.go` instead of keeping it in `s3.go`, and do the same for s3Session?  Semantically it feels weird to have a `s3OS` inside a `gsOS`, and I think it might be a point of confusion for a new developer who is learning the code.

Reply:
I don't really have a preference about naming here, but `s3OS` could be justified considering Google is using a "S3-compatible API" which is what many object storage providers advertise themselves as supporting.

Reply:
I think is `s3` will be better understood by developers - because it is essentially S3 API, and everyone knows what it is.

Reply:
Another option is to add the `host` and `bucket` fields in the `gsOS` struct instead of embedding `s3OS`. We only use two out of six fields in `s3OS`. The redundancy is minimal, and we wouldn't have to rename anything.

Reply:
For me it would be illogical - we still sharing `s3session` and much code around that, so having `s3OS` and `gsOS` seems logical.